### DISCUSSION TITLE
Filtering by calculated column values within a row?

### DISCUSSION THREAD

**Discussion by ksaadDE** [diesel-rs/diesel#3947] [Q&A]
I just stumbled over a interesting issue and use-case, which I'd like to share and hear your suggestions how to implement it properly. 

The table has the following pseudo SQL + Diesel structure:
```
struct myObj:
  ...
  a:bool
  b:bool
  c:datetime
```

then I am running that tiny set of funcs doing this:
```rust
use chrono::{Utc, NaiveDateTime};

pub fn get_current_naive_date_time() -> NaiveDateTime
{
    Utc::now().naive_utc()
}

fn __has_c_reached (row: &myObj) -> bool
{
    get_current_naive_date_time() >=  row.c
}

fn __short_before_c_reached (row: &myObj) -> bool
{
    get_current_naive_date_time() >=  row.c - Duration::days(7)
}

// this is the calculation !!
fn __get_row_calcfield_numeric (row: &myObj) -> i32
{
        match (row.a, row.b,  __short_before_c_reached(&row),  __has_c_reached(&row) )
        {
            (_,_,true,false)    => 1,   // __short_before_c_reached (c1)
            (_, _,_,true)       => 6 ,  // __has_c_reached (c2)
            (true, true,_,_)    => 2,   // a && b
            (true, false,_,_)   => 0,   // a && !b
            (false, true,_,_)   => 4,   // !a && b
            (false, false,_,_)  => 5    // !a && !a
        }
}
```

So what I am doing basically is calculating an new i32 out of a,b,c. It shows it is really dependent on the values a,b and c. Of course, I could write that using filters and select. So calculating based on those values, which however does not return me the "calculated" numbers, which I need in the rest of the test project. 

Tiny result table for __get_row_calcfield_numeric (kinda like a truth table just with numeric results)

| A        | B           | C1  | C2  | R  |
| ------------- |:-------------:| -----:|  -----:| -----:|
| 0     | 0 | 0 |   0 |   5 |
| 0     | 0 | 0 |   1 |   6 |
| 0     | 0 | 1 |   1 |   6 |
| 0     | 1 | 1 |   1 |  4 |
| 1    | 1 | 1 |   0|   1 |
| 1    | 1 | 0 |   0 |   2 |
| 1    | 0 | 0 |   0 |   0 |
| 1    | 1 | 1 |   1 |   2 |

As far as I know that would really suck hard when implementing it in SQL. Further implementing it three times at different places is probably not A+ level coding. 

My first thought was to add a new field to the structure and make it nullable, and update it when the app is updating or pulling the rows. The problem is, this leads to untracked changes to a,b,c and an outdated new field. So this is not a solution for me. 

My second thought was to use triggers and procedures, due to SQLite this is not only not feasible, the documentation is also kinda rare in this regard, and I do not want to make a whole research project out of this xD   

Another connected approach using SQL **if-clauses and virt** fields results in something like that: 
- https://stackoverflow.com/questions/75553721/calculate-a-value-with-multiple-columns-in-a-each-record 
- https://stackoverflow.com/questions/30571528/calculated-column-in-sql-with-if-elif-else-condition
```SQL
SELECT 
             CASE 
                  WHEN a = 0 AND b =0
                     THEN '5' 
                  WHEN NOW() => c 
                     THEN '6'
              END as a,b,c
FROM myobjs
```
I do not implement all of it here - just an example.
How could I implement that in diesel.rs without getting headaches ? 

My third thought was to add some custom filter to the query to achieve the same I would on DB level. However, after searching a bit and looking into possible places in the docs, I have absolutely no clue how I could implement that. Another issue is: I need the resulting value, does not seem very feasible.

My fear is keeping it like above, so querying all rows of the table and "post-processing" them being kinda resource exhaustive. Although Rust is really fast. If I have 1000 entries and 10 users I would query them all (1000\*10\*req_amount) and post-process them in the Rust App instead of calculating the entries "on DB level". Though I think there is no other way to achieve it atm. 

Ideally I would have an extra virtual field that I could just query to do a "give me all with calcVal = 2" like
```SQL
SELECT xyz FROM myobjs WHERE calcVal=2
```
just with myjobs.filter(...).select(...) in diesl.rs

In other frameworks we have the possibility to use virtual fields or lets say custom fields, to process and store those specific calculations and apply them in filters e.g.:
```rust
pub fn myfilterfunc (row: &myObj) -> bool
{
    (__get_row_calcfield_numeric (&row)) == 6) // if calcfield_numeric returns 6 -> return true
}

Ok::<Vec<Task>, Error>(
 myjobs.filter(
     deleted.eq(false)
  ).customFilter(myfilterfunc).select(myObj::as_select()).load(con).unwrap()
)
```
This would do the post-processing much easier and at one place using one-time coded funcs. However, this is obviously not the ideal solution either.

Keep in mind it: While could be certainly a business project, as I am pretty sure this is a common use-case when implementing business-logic, this is a playground project - still trying to get used to diesel.rs's workflow.

Any ideas? How would you guys solve this? Has Diesel.rs some utilities or a specific methodology to solve issues like that much easier? I would be wondering if nobody thought about this before, as I have seen that a lot out there.

**Comment by weiznich** [comment]
I'm sorry, but I cannot follow what you exactly want to achieve. It's really hard to follow your question as it mixes quite a lot of stuff. Try to focus on one question at the time and try to provide specific example of the problem you want to solve. Do not try to build a generic example, just use your real code as that usually makes it much easier to follow such questions.

**Reply by ksaadDE** [reply]
Okay first of all, don't just say there's something mixed in, better you actually point out to things that are mixed in. I just show my thought process of possible solutions I came up with. 

So I have two bools and a datetime (the datetime is needed for two "states" -> two additional bools), those bools return depending on what's true and not (see table) a int, so 1-6. I want to filter on db-level and return those values when I query entities. I can do that by first getting the entity from db and running my func on it.  But how can I query based on that in my db to limit the results on db-level?

In other ORMs I would store the results of the statements in a temporary variable or in an extra column (which is usually not used by the ORM).  Another possibility is to just make a fake getter for that entity property and return pre-defined values, which also can be used during db-level filtering (on select). For the fake getter/setters I don't see how this could work with diesel.rs

At best I can just use my match written in Rust on the DB, so I do not have to code everything twice, as I have to do these calculations not only in Rust but for e.g. filtering I would have to do it on the DB level as well, and that sucks. If I do it only in Rust, I will have to first grab everything that is in the DB and then filtering out using that func, which is not very efficient. 

But how could I proceed, to obtain such a functionality using diesel.rs? SQL Case statements? Manual filtering and have to code my conditions twice? Can I just configure the query using filters? What's the best possible way? 

That's what the more generalized example describes. If I would post the entire code and be concrete on it, it would be more confusing (more to read) and not helpful for others.

**Reply by weiznich** [reply]
I just want to clarify one thing: You ask for help here! That means you should try as much specific information for other to help you. By just requesting things you won't receive any help here. So what's about explaining exactly what you've done so for and where exactly you hit a problem instead of asking an open question how one could do some vague operation? Otherwise I can only point you to the documentation, as most of the things you are asking is already explained there.

**Reply by ksaadDE** [reply]
So you saying you "don't understand" the very detailed and specific use-case but you already know "it's documented"? Okay? 
So where can I find e.g. the SQL Case Expression in your docs? I quickly looked it up. It is not there.... but it was in my example code.

If you really want to help, just look at this [Doctrine ORM example](https://stackoverflow.com/questions/59683930/api-platform-filter-with-calculated-column-i-e-geo-distance) that does exactly what I want to do (just without my SQL Case Expression): 
```
protected function filterProperty(string $property, $value, QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, string $operationName = null)
{
    // otherwise filter is applied to order and page as well
    if ($property !== 'distance') {
        return;
    }

    [$lat, $lng, $radius] = explode(',', $value);

    // add distance in DQL based on filter location point
    $rootAlias = $queryBuilder->getRootAliases()[0];
    $queryBuilder
        ->addSelect('(6371000 * acos(cos(radians(' . $lat . ')) * cos(radians('.$rootAlias.'.lat)) * cos(radians('.$rootAlias.'.lng) - radians(' . $lng . ')) + sin(radians(' . $lat . ')) * sin(radians('.$rootAlias.'.lat)))) AS distance')
        ->having('distance <= :radius')
        ->setParameter('radius', $radius)
        ->orderBy('distance')
    ;
}
```

Basically it calculates the radius on select and filters by the result. Can I do this with diesel or not? If it's possible where should I look at? 

I want to do that just with a more complex condition-based result, like I have shown above (look at the match and table) 

Thank you very much in advance.

**Reply by weiznich** [reply]
Yes diesel can do this.  Given that I feel that this question still misses concrete examples of what you've tried so far I won't spend any time to write an example, but you can use functions like `QueryDsl::select` (for selecting) and `QueryDsl::filter` for the filtering part for this. These accept arbitrary DSL expressions, not just columns.

As for the SQL case expression: That's only available on the master branch yet. Alternativly you can always extend the DSL in your crate to provide support for whatever SQL expression you require. (See the "Extending diesel guide" for details).

**Reply by ksaadDE** [reply]
The docs didn't really help me out with this, as it's really hard to find it. However, some other people have posted things on StackOverflow, and they leave [this code](https://stackoverflow.com/questions/66048172/how-do-i-select-a-subset-of-columns-with-diesel-rs)
```
let result = match airports.select(
    (
        icao_code,
        longitude,
        latitude,
        sql::<Double>(
            &format!("(3959.0 * acos(cos(radians({lat})) * cos(radians(latitude)) * cos(radians(longitude) - radians({long})) + sin(radians({lat})) * sin(radians(latitude)))) AS distance", lat=latitude_reference, long=longitude_reference)
        )
    )
).load::<AirportByDistance>(database_connection)
{
    Ok(result) => result,
    Err(error) => {
        error!("{:?}", error);
        return Err(());
    }
};

for airport in result {
    info!(
        "AIRPORT: {} has {}nm distance",
        airport.icao_code, airport.distance
    );
}
```

The **critical part** is how to implement the conditions that I have shown above....

So for example I need to return and filter by 1-6. If I do not have any SQL Case Expressions (yet). I probably would need to either:
- do it via a boolean expression but those can basically only return 0 or 1
- do it via some crazy math func (equation) that returns my numbers that I need, if conditions are met
- or I could try to use IIF statements, if they are available

Hmm, so for the easy route... is `SELECT IIF(1 < 2, 'Yes', 'No' ) result;` implemented in Diesel.rs? I only find: https://docs.diesel.rs/master/syn/token/struct.If.html

So my question went also along the lines, if that is a way you would recommend or does the ORM provide some shortcut for this? Such shortcuts are always appreciated! 👍  

That's basically all that I wanted :D Some Idea how would you implement it, because I could try to hack it and you would scream the next time I ask for some nudges and help :D

**Reply by ksaadDE** [reply]
Oh and as for Extending Diesel, could I leverage `!sql_function` for this? https://diesel.rs/guides/extending-diesel.html

```rust
use diesel::sql_types::{Nullable, Text};
sql_function! { fn coalesce(x: Nullable<Text>, y: Text) -> Text { my match? }; }

users.select(coalesce(hair_color, "blue"))
```

What would be in here?
```rust
pub fn coalesce<X, Y>(x: X, y: Y) -> coalesce::HelperType<X, Y>
where
    X: AsExpression<Nullable<Text>>,
    Y: AsExpression<Text>,
{
    ... // <--- here? 
}

pub(crate) mod coalesce {
    pub type HelperType<X, Y> = ...;
}
```

Just trying to get an understanding of what I would do there.

**Reply by ksaadDE** [reply]
Another problem is so far, how to do it for SQLite, which is also a requirement that I am facing.

I looked into this
https://docs.diesel.rs/2.1.x/diesel/expression/functions/macro.sql_function.html#custom-aggregate-functions

But how could I modify it to drop my match into that? Or at least build a SQL equivalent of it?

**Reply by weiznich** [reply]
It's really hard for me to follow your unstructured approach here, nevertheless let me try to answer some of your questions:

> So for example I need to return and filter by 1-6. If I do not have any SQL Case Expressions (yet). I probably would need to either:
> 
>    * do it via a boolean expression but those can basically only return 0 or 1
>    * do it via some crazy math func (equation) that returns my numbers that I need, if conditions are met
>    * or I could try to use IIF statements, if they are available
> 
> Hmm, so for the easy route... is SELECT IIF(1 < 2, 'Yes', 'No' ) result; implemented in Diesel.rs? I only find: https://docs.diesel.rs/master/syn/token/struct.If.html
> 
> So my question went also along the lines, if that is a way you would recommend or does the ORM provide some shortcut for this? Such shortcuts are always appreciated! 👍
> 
> That's basically all that I wanted :D Some Idea how would you implement it, because I could try to hack it and you would scream the next time I ask for some nudges and help :D

For the "missing" case expression: As already written you can use the master branch for that or you can implement it in your own code (by essentially just copying the implementation from the master branch). For `IFF` you need to roll your own implementation.

As I still do not understand the problem you are trying to solve I cannot comment on how I would implement it.

> What would be in here?
> ```
> pub fn coalesce<X, Y>(x: X, y: Y) -> coalesce::HelperType<X, Y>
> where
>     X: AsExpression<Nullable<Text>>,
>     Y: AsExpression<Text>,
> {
>    ... // <--- here? 
> }
> 
> pub(crate) mod coalesce {
>     pub type HelperType<X, Y> = ...;
> }
> ```
> Just trying to get an understanding of what I would do there.

That's in implementation detail, that is not relevant for using that macro. The guide explicitly states for that code snippet

>  If we defined this function manually, it would look like this:

and the relevant [documentation](https://docs.diesel.rs/2.1.x/diesel/expression/functions/macro.sql_function.html) states for a similar snippet:

> will generate this code:

So please take the time and read the whole documentation first?

> I looked into this
> https://docs.diesel.rs/2.1.x/diesel/expression/functions/macro.sql_function.html#custom-aggregate-functions
> 
> But how could I modify it to drop my match into that? Or at least build a SQL equivalent of it?

To write this down a bit more explicitly: I'm generally open to help people with their implementation of something if they share a whole project or at least all the relevant code (minimal reproducible example) online. If they cannot share that because it's secret or whatever reason I'm open to discuss a support contract for helping with the implementation, otherwise I only provide general pointers and you are on your own.

**Reply by ksaadDE** [reply]
> I still do not understand the problem you are trying to solve

Please take time to read the problem's description first.

 I want to use multiple columns on db level of a row, to calculate a (condition-based) value like in the Table. I need that value to perform calculations with it and use the variable later on. Like shown in the code

> hard for me to follow your unstructured approach here

I am assuming that listing possible options to tackle a theorized and practically shown specific problem / use-case is a pretty structured approach, even though it deploys explorative means, as no implementation of a solution was selected yet.

> master branch for that or you can implement it in your own code

Yes but the problem is just that hard-coding a upcoming implementation will probably break the software after an cargo dependency update.  Sure removing the custom impl when the new diesel version is released will probably not too difficult but it is additional work. Might it be the case that it's already released on nightly?

> relevant [documentation](https://docs.diesel.rs/2.1.x/diesel/expression/functions/macro.sql_function.html) states for a similar snippet

Yes but there is no working example that shows how to actually use or implement it, which is exactly what you are complaining about (and why I actually put that aside). 

> So please take the time and read the whole documentation first?

You read the whole documentation for every tool you use? I am pretty sure you are much more selective than me when it comes to that. I am really surprised that you still do not see the flaw in your thinking.

BTW: I found another possible solution using XOR-IF, so I would just need to write an regular query using diesel and still obtain the same result. But that way it's still very weird and probably hard to maintain, especially with Diesel.

**Reply by weiznich** [reply]
Ok, as it seems to be still not clear let me write that down explicitly:

I typically response spend a similar amount of effort for answering a question that the person that ask the question spend. So if you do not provide code, I will also not provide code, especially not for open end questions like this ones. I just don't have the time for that in my free time. If you really need a answer to such question consider a asking for a payed consulting instead.

> Yes but there is no working example that shows how to actually use or implement it, which is exactly what you are complaining about (and why I actually put that aside).

I nee 6 complete examples, including usage in the linked documentation, so if you don't brother reading it I really cannot help you. 

> You read the whole documentation for every tool you use? I am pretty sure you are much more selective than me when it comes to that. I am really surprised that you still do not see the flaw in your thinking

Yes, I normally read the documentation first. I also indicate which parts of the documentation sound helpful for my problem and I include information about what I've already tried so far if I ask question.

Anyway consider that my last answer here, until something substantially changes in your approach on asking such questions.

**Reply by ksaadDE** [reply]
Eh, interesting reaction! How does it feel when being faced with oneself? 

> spend a similar amount of effort for answering a question

Clearly not, please look at my initial question and take a step back.

> I also indicate which parts of the documentation sound helpful for my problem and I include information about what I've already tried so far if I ask question

Thank you for being honest! So why are you trying to tell me to read it full, when you are not doing it either? We both are doing "selective reading" for efficiency reasons and if we can not find what we need, we'd rather say that the docs are troublesome. 

> if you do not provide code

![grafik](https://github.com/diesel-rs/diesel/assets/37943746/7003e4bd-253b-46cd-864f-b36c07467faa)


> I see 6 complete examples

![grafik](https://github.com/diesel-rs/diesel/assets/37943746/aa433a87-96a4-4059-93cd-74bd866a39f2)

Very complete! ;) 

> payed consulting instead

Why should I? You do not provide reasons to hire or pay you. First of all it is still my hobby project for getting used to Diesel, trying to see if it is worth using in production, which I am questioning for now. Your continuous insults are a no-no for business. Further you are not showing excellency, instead you are just reverting to insults and "read the docs" comments. Last but not least you are ignoring the effort that has been put into documenting the problem, use-case and my structured approach to it.

Are you living the community spirit or are you just trying to pressure us into your consultancy service? 🤔

**Reply by ksaadDE** [reply]
I actually expected some indication of what direction I could go or what you are absolutely not recommending. I mean you are using Diesel yourself, you probably have solved that issue yourself multiple times, as it's really basic and common.

**Reply by weiznich** [reply]
I recommend that you take a step back and reconsider your tone. After that reread my response again, I already provided these directions by linking the relevant documentation.

Also a word of warning: If you continue this demanding tone here and continue to insult me I will just ban you from interacting with this project. Consider this as last warning.

**Reply by ksaadDE** [reply]
@weiznich If you accuse me of having a demanding tone, please prove it. In my opinion I am very soft and friendly towards you, despite your reply was just "read the docs", "I do not understand", and "buy my service [if you are too stupid to fix it alone]". Then I tried to kindly explain why I am not willing to give you my money and pointing out the community spirit. Yet you threatening me again. It looks like you are arrogant, aggressive, degrading and threatening when confronted with valid and sound critique. I am surprised that you did not at the beginning because your tone was always very rude and very unwelcoming. 

Just let me know if you are ready to focus the initial question instead of your problems.

**Reply by weiznich** [reply]
I'm not interested in continuing this discussion, as it's not productive. I will lock this thread now. Feel free to open a new discussion if you feel that your question is not answered yet, but keep in mind to include at least the following information:

* A minimal reproducible example of the code in question
* A list of documentation you read
* General information about your setup (database system, operating system, rust version, …)
